std/node/tty.ts

```diff
@@ -0,0 +1,295 @@
+      default:
+        return COLORS_2;
+    }
+  }
+
+  if (
+    env.NODE_DISABLE_COLORS !== undefined ||
+    // See https://no-color.org/
+    env.NO_COLOR !== undefined ||
+    // The "dumb" special terminal, as defined by terminfo, doesn't support
+    // ANSI color control codes.
+    // See https://invisible-island.net/ncurses/terminfo.ti.html#toc-_Specials
+    env.TERM === "dumb"
+  ) {
+    return COLORS_2;
+  }
+
+  if (nodeProcess.platform === "win32") {
+    // Lazy load for startup performance.
+    if (OSRelease === undefined) {
+      const { release } = nodeOs;
+      OSRelease = release().split(".");
+    }
+    // Windows 10 build 10586 is the first Windows release that supports 256
+    // colors. Windows 10 build 14931 is the first release that supports
+    // 16m/TrueColor.
+    if (+OSRelease[0] >= 10) {
+      const build = +OSRelease[2];
+      if (build >= 14931) {
+        return COLORS_16M;
+      }
+      if (build >= 10586) {
+        return COLORS_256;
+      }
+    }
+
+    return COLORS_16;
+  }
+
+  if (env.TMUX) {
+    return COLORS_256;
+  }
+
+  if (env.CI) {
+    if (
+      "TRAVIS" in env || "CIRCLECI" in env || "APPVEYOR" in env ||
+      "GITLAB_CI" in env || env.CI_NAME === "codeship"
+    ) {
+      return COLORS_256;
+    }
+    return COLORS_2;
+  }
+
+  if ("TEAMCITY_VERSION" in env) {
+    return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env.TEAMCITY_VERSION)
+      ? COLORS_16
+      : COLORS_2;
+  }
+
+  switch (env.TERM_PROGRAM) {
+    case "iTerm.app":
+      if (
+        !env.TERM_PROGRAM_VERSION ||
+        /^[0-2]\./.test(env.TERM_PROGRAM_VERSION)
+      ) {
+        return COLORS_256;
+      }
+      return COLORS_16M;
+    case "HyperTerm":
+    case "MacTerm":
+      return COLORS_16M;
+    case "Apple_Terminal":
+      return COLORS_256;
+  }
+
+  if (env.COLORTERM === "truecolor" || env.COLORTERM === "24bit") {
+    return COLORS_16M;
+  }
+
+  if (env.TERM) {
+    if (/^xterm-256/.test(env.TERM)) {
+      return COLORS_256;
+    }
+
+    const termEnv = env.TERM.toLowerCase() as keyof TTY_ENVS;
+
+    if (TERM_ENVS[termEnv]) {
+      return TERM_ENVS[termEnv];
+    }
+    for (const term of TERM_ENVS_REG_EXP) {
+      if (term.test(termEnv)) {
+        return COLORS_16;
+      }
+    }
+  }
+  // Move 16 color COLORTERM below 16m and 256
+  if (env.COLORTERM) {
+    return COLORS_16;
+  }
+  return COLORS_2;
+}
+
+export function hasColors(count: number, env: Env) {
+  if (
+    env === undefined &&
+    (count === undefined || (typeof count === "object" && count !== null))
+  ) {
+    env = count;
+    count = 16;
+  } else {
+    if (typeof count !== "number") {
+      throw new ERR_INVALID_ARG_TYPE("count", "number", count);
+    }
+    if (count < 2) {
+      throw new ERR_OUT_OF_RANGE("count", ">= 2", count);
+    }
+  }
+  return count <= 2 ** getColorDepth(env);
+}
+
+export function isatty(fd: number) {
+  // TODO(jopemachine): to be implemented
+  notImplemented();
+}
```

Comment:
It would be really nice to just support stdin / stdout / stderr here as that would cover the majority of use cases for nice terminal output (basically supports-color doing `isatty(1)`

```suggestion
export function isatty(fd: number) {
  switch (fd) {
    case 1: return isStdinTTY();
    case 2: return isStdoutTTY();
    case 3: return isStderrTTY();
  }
  notImplemented();
}
```

where on the Rust side it should just be like https://github.com/denoland/deno/blob/master/runtime/permissions.rs#L605.

Reply:
Thanks you for your reviewing.

May I ask my understanding is correct?

```
function isStdinTTY() {
  return nodeProcess.stdin.isTTY;
}

function isStdoutTTY() {
  return nodeProcess.stdout.isTTY;
}

function isStderrTTY() {
  return nodeProcess.stderr.isTTY;
}

export function isatty(fd: number) {
  switch (fd) {
    case 1: return isStdinTTY();
    case 2: return isStdoutTTY();
    case 3: return isStderrTTY();
  }

  return Deno.isatty(fd);
}
```

Reply:
Just been doing some further digging and this might be as an example to follow - https://github.com/denoland/deno/blob/master/runtime/js/40_tty.js#L11.

Reply:
> Just been doing some further digging and this might be as an example to follow - https://github.com/denoland/deno/blob/master/runtime/js/40_tty.js#L11.

You mean to implement new isatty function using the new rust function, not using Deno.isatty (it seems that maybe it could use 40_tty.js)?

Excusing for my poor understanding, May I ask you why we should do that?

Reply:
Ah, I didn't know there was a `Deno.isatty`! Ok I've corrected the suggestion to a proposed exact implementation then.